modules/seqexec/server/src/main/scala/seqexec/server/ObserveActions.scala

```diff
@@ -4,129 +4,201 @@
-    systems.odb.datasetStart(obsId, dataId, imageFileId)
-      .ensure(SeqexecFailure.Unexpected("Unable to send DataStart message to ODB."))(identity)
+  private def info[F[_]: Logger](msg: => String): F[Unit] = Logger[F].info(msg)
+
+  /**
+    * Actions to perform when an observe is aborted
+    */
+  def abortTail[F[_]: MonadError[?[_], Throwable]](
+    systems:     Systems[F],
+    obsId:       Observation.Id,
+    imageFileId: ImageFileId
+  ): F[Result[F]] =
+    systems.odb
+      .obsAbort(obsId, imageFileId)
+      .ensure(
+        SeqexecFailure
+          .Unexpected("Unable to send ObservationAborted message to ODB.")
+      )(identity) *>
+      MonadError[F, Throwable].raiseError(
+        SeqexecFailure
+          .Execution(s"Observation ${obsId.format} aborted by user.")
+      )
+
+  /**
+    * Send the datasetStart command to the odb
+    */
+  private def sendDataStart[F[_]: MonadError[?[_], Throwable]](
+    systems:     Systems[F],
+    obsId:       Observation.Id,
+    imageFileId: ImageFileId,
+    dataId:      String
+  ): F[Unit] =
+    systems.odb
+      .datasetStart(obsId, dataId, imageFileId)
+      .ensure(
+        SeqexecFailure.Unexpected("Unable to send DataStart message to ODB.")
+      )(identity)
       .void
 
-  private def sendDataEnd[F[_]: MonadError[?[_], Throwable]](systems: Systems[F], obsId: Observation.Id, imageFileId: ImageFileId, dataId: String): F[Unit] =
-    systems.odb.datasetComplete(obsId, dataId, imageFileId)
-      .ensure(SeqexecFailure.Unexpected("Unable to send DataEnd message to ODB."))(identity)
+  /**
+    * Send the datasetEnd command to the odb
+    */
+  private def sendDataEnd[F[_]: MonadError[?[_], Throwable]](
+    systems:     Systems[F],
+    obsId:       Observation.Id,
+    imageFileId: ImageFileId,
+    dataId:      String
+  ): F[Unit] =
+    systems.odb
+      .datasetComplete(obsId, dataId, imageFileId)
+      .ensure(
+        SeqexecFailure.Unexpected("Unable to send DataEnd message to ODB.")
+      )(identity)
       .void
 
-  def observe[F[_]: Sync: Concurrent](
-    systems: Systems[F],
-    config: Config,
-    obsId: Observation.Id,
-    inst: InstrumentSystem[F],
-    otherSys: List[System[F]],
-    headers: Reader[HeaderExtraData, List[Header[F]]])(
-      ctx: HeaderExtraData): Stream[F, Result[F]]
-  = {
-    def dataId: F[String] =
-      EitherT.fromEither[F](
-        config.extractAs[String](OBSERVE_KEY / DATA_LABEL_PROP)
+  /**
+    * Standard progress stream for an observation
+    */
+  def observationProgressStream[F[_]](
+    env: ObserveEnvironment[F]
+  ): Stream[F, Result[F]] =
+    for {
+      ot <- Stream.eval(env.inst.calcObserveTime(env.config))
+      pr <- env.inst.observeProgress(ot, ElapsedTime(0.0.seconds))
+    } yield Result.Partial(pr)
+
+  /**
+    * Tell each subsystem that an observe will start
+    */
+  def notifyObserveStart[F[_]: Applicative](
```

Comment:
Many methods used internally in `observed` are now publicly exposed to let `InstrumentAction` instance use them